### PR TITLE
feat(assertions): `stringLikeRegexp()` matcher

### PR DESCRIPTION
**Pull request by otaviomacedo**
The logic was partly copied from the function with the same name from `assert-internal`, but the implementation was changed to use [minimatch](https://github.com/isaacs/minimatch). This way, we can have a well-known and stable contract for the matcher.

Closes https://github.com/aws/aws-cdk/issues/18051.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
Can you compare with this implementation and see what's different? Niranjan reimplemented it for Pipelines, I don't know if he changed anything:

https://github.com/aws/aws-cdk/blob/604564caa10b5e4b4708a5df8215e213c865dcb6/packages/%40aws-cdk/pipelines/test/testhelpers/matchers.ts#L20-L43

File: packages/@aws-cdk/assertions/README.md [hunk 1]
```diff
@@ -299,6 +299,35 @@ target array. Out of order will be recorded as a match failure.
...
+// }
+
+// The following will NOT throw an assertion error
+template.hasResourceProperties('Foo::Bar', {
+  Fred: Match.stringLike('.*includeHeaders = true.*'),
+});
+
+// The following will throw an assertion error
+template.hasResourceProperties('Foo::Bar', {
+  Fred: Match.stringLike('.*includeHeaders = false.*'),
```
Comment (rix0rrr) [inline packages/@aws-cdk/assertions/README.md:327]: `.*` is not necessary if the pattern doesn't start with a `^`. 

Potentially there is a use for an option which says `{ fullString: true }`, which automatically adds a `^` at the start and a `$` at the end.

**Review by rix0rrr — APPROVED** [review]

File: packages/@aws-cdk/assertions/lib/match.ts [hunk 2]
```diff
@@ -378,3 +385,38 @@ class AnyMatch extends Matcher {
...
+
+class StringLikeRegexpMatch extends Matcher {
+  constructor(
+    public readonly name: string,
+    private readonly pattern: string) {
+
+    super();
+  }
+
+
```
Comment (kaizen3031593) [inline packages/@aws-cdk/assertions/lib/match.ts:397]: extra line?

**Review by kaizen3031593 — APPROVED** [review]
I approve as well :). Thanks for this!

File: packages/@aws-cdk/assertions/lib/template.ts [hunk 3]
```diff
@@ -205,7 +205,11 @@ function toTemplate(stack: Stack): any {
   if (!Stage.isStage(root)) {
     throw new Error('unexpected: all stacks must be part of a Stage or an App');
   }
-  const assembly = root.synth();
+
+  // to support incremental assertions (i.e. "expect(stack).toNotContainSomething(); doSomething(); expect(stack).toContainSomthing()")
```
[the remark this answers is not in the document]
Comment (otaviomacedo) [reply packages/@aws-cdk/assertions/lib/template.ts:209]: In `aws-apigateway` there is a test in which we create a stack, do some assertions, then invalidate a deployment and do more assertions on the same stack to make sure that the old deployment is no longer there.

**Pull request merged**